### PR TITLE
sGho Updates

### PR DESCRIPTION
**Pull request by efecarranza**
Replaces #20 

Builds on top of: https://github.com/aave-dao/gho-origin/pull/5 sGho
I'd make both of these PRs target an sgho side-branch rather than main, merge that one first, then see the changes here before going to main. Have a quick look back by Certora potentially.

Separates tests into different files to make more manageable.
Addresses some style issues.

### CODE REVIEW

File: src/contracts/sgho/sGHO.sol [hunk 1]
```diff
@@ -74,93 +78,34 @@ contract sGHO is
 
   /**
    * @notice Initializer for the sGHO vault.
-   * @param gho_       Address of the underlying GHO token.
-   * @param supplyCap_ The total supply cap for the vault.
-   * @param executor_  The address that will be granted the DEFAULT_ADMIN_ROLE.
+   * @param gho       Address of the underlying GHO token.
+   * @param initialSupplyCap The total supply cap for the vault.
+   * @param executor  The address that will be granted the DEFAULT_ADMIN_ROLE.
```
Comment (pavelvm5) [inline src/contracts/sgho/sGHO.sol]:
```suggestion
   * @param gho Address of the underlying GHO token.
   * @param initialSupplyCap The total supply cap for the vault.
   * @param executor The address that will be granted the DEFAULT_ADMIN_ROLE.
```

File: src/contracts/sgho/sGHO.sol [hunk 2]
```diff
@@ -50,20 +51,23 @@ contract sGHO is
...
+  bytes32 public constant FUNDS_ADMIN_ROLE = 'FUNDS_ADMIN_ROLE';
+
+  /// @inheritdoc IsGho
+  bytes32 public constant PAUSE_GUARDIAN_ROLE = 'PAUSE_GUARDIAN_ROLE';
+
+  /// @inheritdoc IsGho
+  bytes32 public constant TOKEN_RESCUER_ROLE = 'TOKEN_RESCUER_ROLE';
+
+  /// @inheritdoc IsGho
+  bytes32 public constant YIELD_MANAGER_ROLE = 'YIELD_MANAGER_ROLE';
```
Comment (pavelvm5) [inline src/contracts/sgho/sGHO.sol]:
```suggestion
  /// @inheritdoc IsGho
  bytes32 public constant FUNDS_ADMIN_ROLE = keccak256('FUNDS_ADMIN_ROLE)';

  /// @inheritdoc IsGho
  bytes32 public constant PAUSE_GUARDIAN_ROLE = keccak256('PAUSE_GUARDIAN_ROLE');

  /// @inheritdoc IsGho
  bytes32 public constant TOKEN_RESCUER_ROLE = keccak256('TOKEN_RESCUER_ROLE');

  /// @inheritdoc IsGho
  bytes32 public constant YIELD_MANAGER_ROLE = keccak256('YIELD_MANAGER_ROLE');
```

Should be checked in tests too.

File: src/contracts/sgho/sGHO.sol [hunk 3]
```diff
@@ -327,54 +346,12 @@ contract sGHO is
...
-
-  /**
-   * @inheritdoc IsGHO
-   */
-  function ratePerSecond() public view returns (uint96) {
-    return _getsGHOStorage().ratePerSecond;
+  function _getsGhoStorage() private pure returns (sGhoStorage storage $) {
+    assembly {
+      $.slot := sGhoStorageLocation
+    }
```
Comment (pavelvm5) [inline src/contracts/sgho/sGHO.sol]: This is for sure should be moved on top, as it was before, cause this private function of getting storage always placed after storage constant link in all our contracts.

**Review by pavelvm5 — CHANGES_REQUESTED** [review]

**Review by pavelvm5 — APPROVED** [review]

File: src/contracts/sgho/interfaces/IsGHO.sol [hunk 4]
```diff
@@ -56,59 +47,71 @@ interface IsGHO {
...
   /**
-   * @notice Returns the address of the GHO token used as the underlying asset in the vault.
-   * @return The address of the GHO token.
+   * @notice Deposits GHO into the vault using permit and mints sGHO shares to the receiver.
+   * @dev This function allows users to deposit GHO without requiring a separate approve transaction.
+   * The permit is used to approve the vault to spend the user's GHO tokens.
+   * The yield index is updated before the deposit to ensure correct share calculation.
+   * @param assets The amount of GHO to deposit.
+   * @param receiver The address that will receive the sGHO shares.
+   * @param deadline Must be a timestamp in the future.
```
Comment (Kogaroshi) [inline src/contracts/sgho/interfaces/IsGHO.sol]: could explain better that deadline is the max timestamp at which intent can be executed/signature is valid.

File: src/contracts/sgho/interfaces/IsGHO.sol [hunk 5]
```diff
@@ -56,59 +47,71 @@ interface IsGHO {
...
+   * @param newRate The new target rate in basis points (e.g., 1000 for 10%).
    */
-  function yieldIndex() external view returns (uint176);
+  function setTargetRate(uint16 newRate) external;
 
   /**
-   * @notice Returns the current target annual percentage rate (APR) for yield generation.
-   * @dev The rate is expressed in basis points (1% = 100).
-   * @return The target rate in basis points.
+   * @notice Sets the supply cap for the vault.
```
Comment (Kogaroshi) [inline src/contracts/sgho/interfaces/IsGHO.sol:89]: add here or in @dev comment is cap is in asset terms or unit terms

File: src/contracts/sgho/interfaces/IsGHO.sol [hunk 6]
```diff
@@ -117,50 +120,42 @@ interface IsGHO {
...
    */
-  function pause() external;
+  function lastUpdate() external view returns (uint64);
 
   /**
-   * @notice Unpauses the contract, can be called by `PAUSE_GUARDIAN_ROLE`.
-   * Emits a {Unpaused} event.
+   * @notice Returns the current rate per second for yield generation.
+   * @dev The rate is expressed in basis points (1% = 100).
+   * @return The rate per second multiplied by 10^27.
```
Comment (Kogaroshi) [inline src/contracts/sgho/interfaces/IsGHO.sol]: replace "multiplied by 10^27" by "in RAY" ?

File: src/contracts/sgho/interfaces/IsGHO.sol [hunk 7]
```diff
@@ -117,50 +120,42 @@ interface IsGHO {
...
   /**
-   * @notice Sets the target rate for yield generation.
-   * @dev This function can only be called by an address with the YIELD_MANAGER role.
-   * The new rate must be less than 50% (5000 basis points).
-   * @param newRate The new target rate in basis points (e.g., 1000 for 10%).
+   * @notice Returns the total supply cap of the vault.
+   * @return The total supply cap.
    */
-  function setTargetRate(uint16 newRate) external;
+  function supplyCap() external view returns (uint160);
```
Comment (Kogaroshi) [inline src/contracts/sgho/interfaces/IsGHO.sol:148]: same than https://github.com/aave-dao/gho-origin/pull/22/files#r2753689436

File: src/contracts/sgho/interfaces/IsGHO.sol [hunk 8]
```diff
@@ -117,50 +120,42 @@ interface IsGHO {
...
-   * The permit is used to approve the vault to spend the user's GHO tokens.
-   * The yield index is updated before the deposit to ensure correct share calculation.
-   * @param assets The amount of GHO to deposit.
-   * @param receiver The address that will receive the sGHO shares.
-   * @param deadline Must be a timestamp in the future.
-   * @param sig A `secp256k1` signature params from `msgSender()`.
-   * @return The amount of sGHO shares minted.
+   * @notice Returns the current yield index, representing the accumulated yield.
+   * @dev This index is used to calculate the value of sGHO in terms of GHO.
+   * @return The current yield index.
```
Comment (Kogaroshi) [inline src/contracts/sgho/interfaces/IsGHO.sol:160]: specify the scale used for the index (in RAY)

File: src/contracts/sgho/sGHO.sol [hunk 9]
```diff
@@ -1,21 +1,22 @@
...
+import {IERC20} from 'openzeppelin-contracts/contracts/token/ERC20/IERC20.sol';
+import {IERC4626} from 'openzeppelin-contracts/contracts/interfaces/IERC4626.sol';
+import {IERC20Metadata} from 'openzeppelin-contracts/contracts/token/ERC20/extensions/IERC20Metadata.sol';
 import {ERC4626Upgradeable} from 'openzeppelin-contracts-upgradeable/contracts/token/ERC20/extensions/ERC4626Upgradeable.sol';
 import {ERC20PermitUpgradeable} from 'openzeppelin-contracts-upgradeable/contracts/token/ERC20/extensions/ERC20PermitUpgradeable.sol';
 import {ERC20Upgradeable} from 'openzeppelin-contracts-upgradeable/contracts/token/ERC20/ERC20Upgradeable.sol';
 import {AccessControlUpgradeable} from 'openzeppelin-contracts-upgradeable/contracts/access/AccessControlUpgradeable.sol';
+import {Initializable} from 'openzeppelin-contracts-upgradeable/contracts/proxy/utils/Initializable.sol';
+import {Math} from 'openzeppelin-contracts/contracts/utils/math/Math.sol';
+import {SafeCast} from 'openzeppelin-contracts/contracts/utils/math/SafeCast.sol';
```
Comment (Kogaroshi) [inline src/contracts/sgho/sGHO.sol:14]: why the change of order ?

File: src/contracts/sgho/sGHO.sol [hunk 10]
```diff
@@ -50,20 +51,29 @@ contract sGHO is
   }
 
   // keccak256(abi.encode(uint256(keccak256("gho.storage.sGHO")) - 1)) & ~bytes32(uint256(0xff))
```
Comment (Kogaroshi) [inline src/contracts/sgho/sGHO.sol]: "gho.storage.sGHO" => reflect the contract rename here too

File: src/contracts/sgho/sGHO.sol [hunk 11]
```diff
@@ -74,93 +78,34 @@ contract sGHO is
...
     $.targetRate = 0;
   }
 
-  /**
-   * @notice The receive function is implemented to reject direct Ether transfers to the contract.
-   * @dev sGHO does not handle ETH directly. All deposits must be made in the GHO token.
-   */
-  receive() external payable {
-    revert NoEthAllowed();
-  }
```
[the remark this answers is not in the document]
Comment (Kogaroshi) [reply src/contracts/sgho/sGHO.sol:110]: No need to bring it back, imo `receive` is useless here

File: tests/unit/TestSGhoBase.t.sol [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+  function setUp() public virtual {
+    initTestEnvironment(false); // Use TestnetProcedures setup
+
+    // Users
+    user1PrivateKey = 0xB0B;
+    user1 = vm.addr(user1PrivateKey);
+    user2 = vm.addr(0xCAFE);
+    Admin = vm.addr(0x1234); // proxy admin address
+    yManager = vm.addr(0xDEAD); // Yield manager address
+    fundsAdmin = vm.addr(0xA11D); // Funds admin address
```
Comment (Kogaroshi) [inline tests/unit/TestSGhoBase.t.sol]: Consider using `makeAddr` & `makeAddrAndKey` from Foundry here : https://getfoundry.sh/reference/forge-std/make-addr#makeaddr
example : `(user1, user1PrivateKey) = makeAddrAndKey('user1')`

File: tests/unit/TestSGhoBase.t.sol [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+
+    // Deploy Mocks & sGho
+    gho = new TestnetERC20('Mock GHO', 'GHO', 18, poolAdmin);
+
+    // Deploy sGho implementation and proxy
+    address sghoImpl = address(new sGho());
+    sgho = sGho(
+      payable(
```
Comment (Kogaroshi) [inline tests/unit/TestSGhoBase.t.sol]: is `payable` still necessary here ?

File: tests/unit/TestSGhoBase.t.sol [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+    uint256 prevYieldIndex,
+    uint16 targetRate,
+    uint256 timeSinceLastUpdate
+  ) internal pure returns (uint256) {
+    if (targetRate == 0 || timeSinceLastUpdate == 0) return prevYieldIndex;
+
+    // Convert targetRate from basis points to ray
+    uint256 annualRateRay = (uint256(targetRate) * RAY) / 10000;
+    // Calculate the rate per second (new contract logic)
+    uint256 ratePerSecond = (annualRateRay * RAY) / 365 days;
```
Comment (Kogaroshi) [inline tests/unit/TestSGhoBase.t.sol]: `ratePerSecondRay` for consistency here

File: tests/unit/TestSGhoBase.t.sol [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+    address spender,
+    uint256 value,
+    uint256 nonce,
+    uint256 deadline,
+    uint256 privateKey
+  ) internal view returns (uint8 v, bytes32 r, bytes32 s) {
+    bytes32 PERMIT_TYPEHASH = keccak256(
+      'Permit(address owner,address spender,uint256 value,uint256 nonce,uint256 deadline)'
+    );
+    bytes32 structHash = keccak256(
```
Comment (Kogaroshi) [inline tests/unit/TestSGhoBase.t.sol]: consider using Foundry's cheatcodes here to validate types and EIP-712 implementation in sGho : 
https://getfoundry.sh/guides/eip712

File: tests/unit/TestSGhoPause.t.sol [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+  // ========================================
+  // PAUSABILITY TESTS
+  // ========================================
+
+  function test_pausability_deposit_withdraw() external {
+    address pauseGuardian = vm.addr(0xBAD);
+    sgho.grantRole(sgho.PAUSE_GUARDIAN_ROLE(), pauseGuardian);
+
+    // Pause the contract
+    vm.startPrank(pauseGuardian);
```
Comment (Kogaroshi) [inline tests/unit/TestSGhoPause.t.sol]: consider using only vm.prank() here instead of start/stop, in the cases where only 1 call is made during the prank, to simplify the code here

File: tests/unit/TestSGhoPrecision.t.sol [hunk 17]
```diff
@@ -20,16 +19,16 @@ import {Math} from 'openzeppelin-contracts/contracts/utils/math/Math.sol';
  * 4. Edge Case Tests - Validate behavior with extreme values
  * 5. Actual Measurement Tests - Output real precision loss values for analysis
  */
-contract sGhoPrecisionTest is TestnetProcedures {
+contract TestSGhoPrecision is TestnetProcedures {
```
Comment (Kogaroshi) [inline tests/unit/TestSGhoPrecision.t.sol:22]: could inherit `TestSGhoBase` instead

File: tests/unit/TestSGhoRatesAndCaps.t.sol [hunk 18]
```diff
@@ -0,0 +1,176 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import './TestSGhoBase.t.sol';
+
+contract TestSGhoRatesAndCaps is TestSGhoBase {
```
Comment (Kogaroshi) [inline tests/unit/TestSGhoRatesAndCaps.t.sol]: `TestSGhoConfig` ?

File: tests/unit/TestSGhoRatesAndCaps.t.sol [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+    vm.expectRevert(IsGho.RateMustBeLessThanMaxRate.selector);
+    sgho.setTargetRate(newRate);
+    vm.stopPrank();
+  }
+
+  // ========================================
+  // SUPPLY CAP & LIMITS TESTS
+  // ========================================
+
+  function test_revert_deposit_exceedsCap() external {
```
Comment (Kogaroshi) [inline tests/unit/TestSGhoRatesAndCaps.t.sol]: Those tests should be with the other deposit/mint tests & such, as reverting cases.
Only keep the setters and checks in this test file.

File: tests/unit/TestSGhoRatesAndCaps.t.sol [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+      expectedMaxMint,
+      'maxMint should be remaining capacity in shares'
+    );
+  }
+
+  // ========================================
+  // EVENT TESTS
+  // ========================================
+
+  function test_ExchangeRateUpdateEvent_basic() external {
```
Comment (Kogaroshi) [inline tests/unit/TestSGhoRatesAndCaps.t.sol]: move this test to `TestSGhoYield` ?

**Comment by Kogaroshi** [comment]
Pls also fix the import for CI run

File: src/contracts/sgho/interfaces/IsGHO.sol (hunk 5 above)
Comment (Kogaroshi) [reply src/contracts/sgho/interfaces/IsGHO.sol:89]: fixed in d18cb8c0c794ee2ef689f6bde9366717b9da721f

File: src/contracts/sgho/interfaces/IsGHO.sol (hunk 7 above)
Comment (Kogaroshi) [reply src/contracts/sgho/interfaces/IsGHO.sol:148]: fixed in d18cb8c0c794ee2ef689f6bde9366717b9da721f

File: src/contracts/sgho/interfaces/IsGHO.sol (hunk 4 above)
Comment (Kogaroshi) [reply src/contracts/sgho/interfaces/IsGHO.sol]: fixed in d18cb8c0c794ee2ef689f6bde9366717b9da721f

File: src/contracts/sgho/interfaces/IsGHO.sol (hunk 6 above)
Comment (Kogaroshi) [reply src/contracts/sgho/interfaces/IsGHO.sol]: fixed in d18cb8c0c794ee2ef689f6bde9366717b9da721f

File: src/contracts/sgho/interfaces/IsGHO.sol (hunk 8 above)
Comment (Kogaroshi) [reply src/contracts/sgho/interfaces/IsGHO.sol:160]: fixed in d18cb8c0c794ee2ef689f6bde9366717b9da721f

File: src/contracts/sgho/sGHO.sol (hunk 10 above)
Comment (Kogaroshi) [reply src/contracts/sgho/sGHO.sol]: fixed in d18cb8c0c794ee2ef689f6bde9366717b9da721f

File: tests/unit/TestSGhoBase.t.sol (hunk 12 above)
Comment (Kogaroshi) [reply tests/unit/TestSGhoBase.t.sol]: fixed in d18cb8c0c794ee2ef689f6bde9366717b9da721f

File: tests/unit/TestSGhoBase.t.sol (hunk 13 above)
Comment (Kogaroshi) [reply tests/unit/TestSGhoBase.t.sol]: fixed in d18cb8c0c794ee2ef689f6bde9366717b9da721f

File: tests/unit/TestSGhoBase.t.sol (hunk 14 above)
Comment (Kogaroshi) [reply tests/unit/TestSGhoBase.t.sol]: fixed in d18cb8c0c794ee2ef689f6bde9366717b9da721f

File: tests/unit/TestSGhoBase.t.sol (hunk 15 above)
Comment (Kogaroshi) [reply tests/unit/TestSGhoBase.t.sol]: Changes made in d18cb8c0c794ee2ef689f6bde9366717b9da721f

Could be further improved by using json-binding to also verify the type returned by the contract, but current version is enough for tests as well.

File: tests/unit/TestSGhoPause.t.sol (hunk 16 above)
Comment (Kogaroshi) [reply tests/unit/TestSGhoPause.t.sol]: fixed in d18cb8c0c794ee2ef689f6bde9366717b9da721f

File: tests/unit/TestSGhoRatesAndCaps.t.sol (hunk 18 above)
Comment (Kogaroshi) [reply tests/unit/TestSGhoRatesAndCaps.t.sol]: fixed in d18cb8c0c794ee2ef689f6bde9366717b9da721f

File: tests/unit/TestSGhoRatesAndCaps.t.sol (hunk 19 above)
Comment (Kogaroshi) [reply tests/unit/TestSGhoRatesAndCaps.t.sol]: fixed in d18cb8c0c794ee2ef689f6bde9366717b9da721f

File: tests/unit/TestSGhoRatesAndCaps.t.sol (hunk 20 above)
Comment (Kogaroshi) [reply tests/unit/TestSGhoRatesAndCaps.t.sol]: fixed in d18cb8c0c794ee2ef689f6bde9366717b9da721f

File: tests/unit/TestSGhoInitialization.t.sol [hunk 21]
```diff
@@ -0,0 +1,205 @@
...
+      address(this),
+      abi.encodeWithSelector(
+        sGho.initialize.selector,
+        address(gho),
+        SUPPLY_CAP,
+        address(this) // executor
+      )
+    );
+
+    sGho newSgho = sGho(payable(address(proxy)));
```
Comment (Kogaroshi) [inline tests/unit/TestSGhoInitialization.t.sol]: remove `payable` here too

File: tests/unit/TestSGhoPrecision.t.sol [hunk 22]
```diff
@@ -65,14 +64,14 @@ contract sGhoPrecisionTest is TestnetProcedures {
       ''
     );
 
-    sgho = sGHO(payable(address(proxy)));
+    sgho = sGho(payable(address(proxy)));
```
Comment (Kogaroshi) [inline tests/unit/TestSGhoPrecision.t.sol]: https://github.com/aave-dao/gho-origin/pull/22#discussion_r2755401776

File: tests/unit/TestSGhoInitialization.t.sol [hunk 23]
```diff
@@ -0,0 +1,205 @@
...
+    assertEq(sgho.totalAssets(), 0, 'Initial totalAssets mismatch');
+    assertEq(sgho.totalSupply(), 0, 'Initial totalSupply mismatch');
+    assertEq(sgho.decimals(), gho.decimals(), 'Decimals mismatch'); // Inherits ERC20 decimals
+  }
+
+  function test_initialization() external {
+    // Deploy a new sGho instance
+    address impl = address(new sGho());
+    sGho newSgho = sGho(
+      payable(
```
Comment (Kogaroshi) [inline tests/unit/TestSGhoInitialization.t.sol]: https://github.com/aave-dao/gho-origin/pull/22#discussion_r2755401776

File: tests/unit/TestSGhoBase.t.sol (hunk 15 above)
Comment (efecarranza) [reply tests/unit/TestSGhoBase.t.sol]: will keep as is.

File: tests/unit/TestSGhoInitialization.t.sol (hunk 21 above)
Comment (Kogaroshi) [reply tests/unit/TestSGhoInitialization.t.sol]: Fixed in 59d99786e70e04bfe17fe0576f02986a318c03d3

**Review by Kogaroshi — APPROVED** [review]
